nodeup/pkg/model/etc_hosts.go

```diff
@@ -30,32 +30,37 @@ var _ fi.ModelBuilder = &EtcHostsBuilder{}
 
 // Build is responsible for configuring the gossip DNS tasks.
 func (b *EtcHostsBuilder) Build(c *fi.ModelBuilderContext) error {
-	useGossip := b.Cluster.IsGossip()
-	if !useGossip {
-		return nil
-	}
 
-	if b.IsMaster {
-		task := &nodetasks.UpdateEtcHostsTask{
-			Name: "control-plane-bootstrap",
-		}
+	task := &nodetasks.UpdateEtcHostsTask{
+		Name: "control-plane-address",
+	}
 
-		if b.Cluster.Spec.MasterInternalName != "" {
+	if b.IsMaster && b.Cluster.IsGossip() {
+		task.Records = append(task.Records, nodetasks.HostRecord{
+			Hostname:  b.Cluster.Spec.MasterInternalName,
+			Addresses: []string{"127.0.0.1"},
+		})
+		if b.Cluster.Spec.MasterPublicName != "" {
 			task.Records = append(task.Records, nodetasks.HostRecord{
-				Hostname:  b.Cluster.Spec.MasterInternalName,
+				Hostname:  b.Cluster.Spec.MasterPublicName,
 				Addresses: []string{"127.0.0.1"},
 			})
 		}
-		if b.Cluster.Spec.MasterPublicName != "" {
+	} else if b.Cluster.UsesNoneDNS() {
+		task.Records = append(task.Records, nodetasks.HostRecord{
+			Hostname:  b.Cluster.Spec.MasterInternalName,
+			Addresses: []string{b.BootConfig.APIServer},
+		})
+		if b.UseKopsControllerForNodeBootstrap() {
 			task.Records = append(task.Records, nodetasks.HostRecord{
-				Hostname:  b.Cluster.Spec.MasterPublicName,
-				Addresses: []string{"127.0.0.1"},
+				Hostname:  "kops-controller.internal." + b.Cluster.Name,
+				Addresses: []string{b.BootConfig.APIServer},
```

Comment:
Perhaps apiserver nodes don't make sense when using none DNS? If that's the case then there should be API validation against the combination.

Reply:
None DNS implies at this time a load balancer or something similar and I assume that it would handle both API servers and kops-controller connections. Probably would be best to keep this part out of this PR.

Reply:
As I understand apiserver nodes, the load balancer for the apiserver would go to the apiserver nodes, not the control plane nodes. You'd need a separate load balancer for kops-controller. @olemarkus ?

Reply:
@johngmyers For kops-controller, we need a new listener + target group. See https://github.com/kubernetes/kops/pull/14452 for details.

Reply:
I also added an integration test for it, so this should make it easier to follow the actual changes from Public DNS to None DNS: https://github.com/kubernetes/kops/pull/14452/commits/062adb078ccb7d9652066f1f70a53a7c8cd220cd.